crates/lib3h/src/engine/ghost_engine.rs

```diff
@@ -12,3 +37,1064 @@ pub type GhostEngineParentWrapper<Core, Engine, EngineError> = GhostParentWrappe
+            network_connections: HashSet::new(),
+            space_gateway_map: HashMap::new(),
+            transport_keys,
+            process_count: 0,
+            client_endpoint: Some(endpoint_parent),
+            lib3h_endpoint: Detach::new(
+                endpoint_self
+                    .as_context_endpoint_builder()
+                    .request_id_prefix(name)
+                    .build(),
+            ),
+        };
+
+        engine.priv_connect_bootstraps()?;
+        Ok(engine)
+    }
+
+    fn priv_connect_bootstraps(&mut self) -> GhostResult<()> {
+        let nodes: Vec<Url> = self.config.bootstrap_nodes.drain(..).collect();
+        for bs in nodes {
+            self.handle_connect(ConnectData {
+                request_id: format!("bootstrap-connect: {}", bs.clone()).to_string(), // fire-and-forget
+                peer_uri: bs,
+                network_id: "".to_string(), // unimplemented
+            })?;
+        }
+        Ok(())
+    }
+
+    pub fn this_space_peer(&mut self, chain_id: ChainId) -> PeerData {
+        trace!("engine.this_space_peer() ...");
+        let space_gateway = self
+            .space_gateway_map
+            .get_mut(&chain_id)
+            .expect("No space at chainId");
+        space_gateway.as_mut().this_peer()
+    }
+}
+
+impl<'engine>
+    GhostActor<
+        Lib3hToClient,
+        Lib3hToClientResponse,
+        ClientToLib3h,
+        ClientToLib3hResponse,
+        Lib3hError,
+    > for GhostEngine<'engine>
+{
+    // START BOILER PLATE--------------------------
+    fn take_parent_endpoint(
+        &mut self,
+    ) -> Option<
+        GhostEndpoint<
+            ClientToLib3h,
+            ClientToLib3hResponse,
+            Lib3hToClient,
+            Lib3hToClientResponse,
+            Lib3hError,
+        >,
+    > {
+        std::mem::replace(&mut self.client_endpoint, None)
+    }
+    // END BOILER PLATE--------------------------
+
+    fn process_concrete(&mut self) -> GhostResult<WorkWasDone> {
+        // START BOILER PLATE--------------------------
+        // always run the endpoint process loop
+        detach_run!(&mut self.lib3h_endpoint, |cs| { cs.process(self) })?;
+        // END BOILER PLATE--------------------------
+
+        // process any messages from the client to us
+        for msg in self.lib3h_endpoint.as_mut().drain_messages() {
+            self.handle_msg_from_client(msg)?;
+        }
+
+        /*
+                let outbox: Vec<Lib3hServerProtocol> = Vec::new();
+                let (net_did_work, mut net_outbox) = self.process_multiplexer()?;
+                outbox.append(&mut net_outbox);
+                // Process the space layer
+                let mut p2p_output = self.process_space_gateways()?;
+                outbox.append(&mut p2p_output);
+                // Hack
+                let (ugly_did_work, mut ugly_outbox) = self.process_ugly();
+                outbox.append(&mut ugly_outbox);
+        */
```

Comment:
Should these be un-commented before merge?

Reply:
they are current subtask next steps, can't be merged as is because they return outbox.  Added a TODO and issue # for each.